baybe/campaign.py

```diff
@@ -222,6 +243,68 @@ def add_measurements(
             numerical_measurements_must_be_within_tolerance,
         )
 
+    def _mark_as_measured(
+        self,
+        measurements: pd.DataFrame,
+        numerical_measurements_must_be_within_tolerance: bool,
+    ) -> None:
+        """Mark the given elements of the space as measured.
+
+        Args:
+            measurements: A dataframe containing parameter configurations to be
+                marked as measured.
+            numerical_measurements_must_be_within_tolerance: See
+                :func:`baybe.utils.dataframe.fuzzy_row_match`.
+        """
+        idxs_matched = fuzzy_row_match(
+            self.searchspace.discrete.exp_rep,
+            measurements,
+            self.parameters,
+            numerical_measurements_must_be_within_tolerance,
+        )
+        self._searchspace_metadata.loc[idxs_matched, _MEASURED] = True
+
+    def toggle_discrete_candidates(
+        self,
+        filter: pd.DataFrame,
+        exclude: bool,
+        anti: bool = False,
+        dry_run: bool = False,
+    ) -> pd.DataFrame:
+        """Un-/exclude certain discrete points from the candidate set.
+
+        Args:
+            filter: A dataframe specifying the filtering mechanism to determine the
+                candidates subset to be un-/excluded. The subset is determined via a
+                join (see ``anti`` argument for details) with the discrete space.
+            exclude: If ``True``, the specified candidates are excluded.
+                If ``False``, the candidates are considered for recommendation.
+            anti: If ``False``, the filter determines the points to be affected (i.e.
+                selection via regular join). If ``True``, the filtering mechanism is
+                inverted in that only the points passing the filter are unaffected (i.e.
+                selection via anti-join).
+            dry_run: If ``True``, the target subset is only extracted but not
+                affected. If ``False``, the candidate set is updated correspondingly.
+                Useful for setting up the correct filtering mechanism.
+
+        Returns:
+            The discrete candidate set passing through the specified filter.
+        """
+        exp_rep = self.searchspace.discrete.exp_rep
+        index_name = exp_rep.index.name
+        points = pd.merge(
+            exp_rep.reset_index(names="_df_index"), filter, how="left", indicator=True
+        ).set_index("_df_index")
+        selector = "left_only" if anti else "both"
+        points = points[points["_merge"] == selector]
```

Comment:
these lines could be combined and the drop could also be made inplace (like the drop below)

Reply:
Good idea with the inplace dropping. Have a look at the new version and let me know if this is what you had in mind